controllers/medusa/medusatask_controller.go

```diff
@@ -0,0 +1,450 @@
+		task.Status.InProgress = append(task.Status.InProgress, pod.Name)
+	}
+
+	if err := r.Status().Patch(ctx, task, patch); err != nil {
+		logger.Error(err, "Failed to patch status")
+		// We received a stale object, requeue for next processing
+		return ctrl.Result{}, err
+	}
+
+	// Invoke the purge operation on all pods, in the background
+	go func() {
+		wg := sync.WaitGroup{}
+
+		// Mutex to prevent concurrent updates to the Status object
+		backupMutex := sync.Mutex{}
+		patch := client.MergeFrom(task.DeepCopy())
+
+		for _, p := range pods {
+			pod := p
+			wg.Add(1)
+			go func() {
+				logger.Info("starting purge", "CassandraPod", pod.Name)
+				succeeded := false
+				var response *medusa.PurgeBackupsResponse
+				response, purgeErr := doPurge(ctx, &pod, r.ClientFactory)
+				if purgeErr == nil {
+					logger.Info("finished purge", "CassandraPod", pod.Name)
+					succeeded = true
+				} else {
+					logger.Error(purgeErr, "purge failed", "CassandraPod", pod.Name)
+				}
+				backupMutex.Lock()
+				defer backupMutex.Unlock()
+				defer wg.Done()
+				task.Status.InProgress = utils.RemoveValue(task.Status.InProgress, pod.Name)
+				if succeeded && response != nil {
+					purgeResponse := medusav1alpha1.TaskResult{
+						PodName:                   pod.Name,
+						NbBackupsPurged:           int(response.NbBackupsPurged),
+						NbObjectsPurged:           int(response.NbObjectsPurged),
+						TotalPurgedSize:           int(response.TotalPurgedSize),
+						TotalObjectsWithinGcGrace: int(response.TotalObjectsWithinGcGrace),
+					}
+					task.Status.Finished = append(task.Status.Finished, purgeResponse)
+				} else {
+					task.Status.Failed = append(task.Status.Failed, pod.Name)
+				}
+			}()
+		}
+		wg.Wait()
+		logger.Info("finished task operations")
+		if err := r.Status().Patch(context.Background(), task, patch); err != nil {
+			logger.Error(err, "failed to patch status", "MedusaTask", fmt.Sprint(task))
+		}
+	}()
+
+	return ctrl.Result{RequeueAfter: r.DefaultDelay}, nil
+}
+
+func (r *MedusaTaskReconciler) prepareRestoreOperation(ctx context.Context, task *medusav1alpha1.MedusaTask, pods []corev1.Pod, logger logr.Logger) (reconcile.Result, error) {
+	logger.Info("Starting prepare restore operations")
+
+	// Set the start time and add the pods to the in progress list
+	patch := client.MergeFrom(task.DeepCopy())
+	task.Status.StartTime = metav1.Now()
+	for _, pod := range pods {
+		task.Status.InProgress = append(task.Status.InProgress, pod.Name)
+	}
+
+	if err := r.Status().Patch(ctx, task, patch); err != nil {
+		logger.Error(err, "Failed to patch status")
+		// We received a stale object, requeue for next processing
+		return ctrl.Result{}, err
+	}
+
+	// Invoke the purge operation on all pods, in the background
+	go func() {
+		wg := sync.WaitGroup{}
+
+		// Mutex to prevent concurrent updates to the Status object
+		backupMutex := sync.Mutex{}
+		patch := client.MergeFrom(task.DeepCopy())
+
+		for _, p := range pods {
```

Comment:
The control flow in this loop in nearly identical to the control flow in `purgeOperation` as well as in `MedusaBackupJobReconciler.Reconcile`. The status updates appear to be the same as well. I realize that in one case we are updating a MedusaTaskStatus and in other case we are updating a MedusaBackupJobStatus. How about moving the common fields into shared, common struct? And maybe define an interface through which the status updates are made. Both MedusaTaskStatus and MedusaBackupJobStatus would implement the interface.

Reply:
I forgot to include my main point 🤦‍♂️ The control logic is duplicated in several places. It would be good to look at deduplicating it. Doing so would handling both MedusaTaskStatus and MedusaBackupJobStatus, hence my suggestions for the common struct.

Reply:
@adejanovski I pushed some initial changes [here](https://github.com/jsanda/k8ssandra-operator/blob/medusa-pod-ops/controllers/medusa/medusatask_controller.go#L217) to illustrate what I had in mind. I introduced a `executePodOperations` which encapsulates all of the control flow and goroutines for making calls against individual pods. Some more work is needed to handle the additional status updates in `purgeOperation`. With the status-related refactoring I mentioned in my earlier [comment](https://github.com/k8ssandra/k8ssandra-operator/pull/454#discussion_r858678073) you should be able reuse this  in the backup controller as well.

To be clear though, my main observation is that the control flow for executing operations against pod is duplicated in multiple places.

Reply:
Cool, I've refactored the code using your experiment code. Haven't pushed it as far as refactoring the status because I'd like to work on using common code between task and backup operations in a subsequent PR.
There's a lot being done here already IMO and we should scope this out.
I'll create a ticket for this if you agree.

Reply:
Sorry for the delayed response. Follow up ticket works for me.

Reply:
Cool, I've create the issue: https://github.com/k8ssandra/k8ssandra-operator/issues/541